### ISSUE TITLE
React Props validation does not work for extended classes

### ISSUE THREAD

**Issue by r-tock** [Microsoft/TypeScript#15858]
We have a generic BaseComponent that does some autobinding and transition blocking. 

If I have a component that extends BaseComponent<Props, State>, the compiler does not recognize error in props validation.

The following component does not throw compiler errors.
```jsx
interface Props {
  notMyField: string,
}

class MyComponent extends BaseComponent<Props, any> {
   render() {
      return <div>{this.props.myField}</div>;
   } 
}
```

```jsx
import * as PropTypes from 'prop-types';
import * as React from 'react';

/**
 * The base component that autobinds all the methods to this.
 */
class BaseComponent<Props, State> extends React.Component<Props, State> {
  static propTypes;
  static contextTypes = {
    router: PropTypes.object,
  };

  props;
  unblock;

  constructor(props: Props) {
    super(props);
    Object.getOwnPropertyNames(Object.getPrototypeOf(this)).forEach((method) => {
      if (typeof this[method] !== 'function') {
        return;
      }
      this[method] = this[method].bind(this);
    });
  }

  componentDidMount() {
    const {
      props: {
        history
      },
      willTransition
    } = this as any;
    // TODO: Figure out why react/prop-types is warning here.
    /* eslint-disable react/prop-types */
    if (history && willTransition) {
      this.unblock = history.block(willTransition);
    }
    /* eslint-enable react/prop-types */
  }

  componentWillUnmount() {
    if (this.unblock) {
      this.unblock();
    }
  }
}

export default BaseComponent;
```

**Comment by mhegazy** [comment]
`BaseComponent` has a declaration for `props` with no type, and hence it gets a type `any`. remove the declaration of `props`. 
```ts
class BaseComponent<Props, State> extends React.Component<Props, State> {
    ...
    // props;
   ...
```

also worth nothing that using `--noImplicitAny` would have flagged these properties as `any` for you.

**Comment by r-tock** [comment]
Makes sense.
